etc/tune.c

```diff
@@ -473,160 +431,40 @@ int main(int argc, char **argv)
-         puts("# Karatsuba multiplication.");
-      }
-      for (x = 8; x < upper_limit_print; x += increment_print) {
-         KARATSUBA_MUL_CUTOFF = INT_MAX;
-         t1 = s_time_mul(x);
-         if ((t1 == 0uLL) || (t1 == UINT64_MAX)) {
-            fprintf(stderr,"Karatsuba multiplication  failed at x = INT_MAX (%s)\n",
-                    (t1 == 0uLL)?"wrong result":"internal error");
-            exit(EXIT_FAILURE);
-         }
-         KARATSUBA_MUL_CUTOFF = x;
-         t2 = s_time_mul(x);
-         if ((t2 == 0uLL) || (t2 == UINT64_MAX)) {
-            fprintf(stderr,"Karatsuba multiplication failed (%s)\n",
-                    (t2 == 0uLL)?"wrong result":"internal error");
-            exit(EXIT_FAILURE);
-         }
-         if (verbose == 1) {
-            printf("%d: %9"PRIu64" %9"PRIu64", %9"PRIi64"\n", x, t1, t2, (int64_t)t2 - (int64_t)t1);
-         }
-         if (t2 < t1) {
-            if (count == s_stabilization_extra) {
-               count = 0;
-               break;
-            } else if (count < s_stabilization_extra) {
-               count++;
-            }
-         } else if (count > 0) {
-            count--;
-         }
-      }
-      KARATSUBA_MUL_CUTOFF = x - s_stabilization_extra * increment_print;
+      s_run("Karatsuba multiplication", s_time_mul, &KARATSUBA_MUL_CUTOFF);
+      updated.kmul = KARATSUBA_MUL_CUTOFF;
+      KARATSUBA_MUL_CUTOFF = INT_MAX;
 #endif
 #ifdef BN_S_MP_KARATSUBA_SQR_C
-      if ((verbose == 1) || (testmode == 1)) {
-         puts("# Karatsuba squaring.");
-      }
-      for (x = 8; x < upper_limit_print; x += increment_print) {
-         KARATSUBA_SQR_CUTOFF = INT_MAX;
-         t1 = s_time_sqr(x);
-         if ((t1 == 0uLL) || (t1 == UINT64_MAX)) {
-            fprintf(stderr,"Karatsuba squaring failed at x = INT_MAX (%s)\n",
-                    (t1 == 0uLL)?"wrong result":"internal error");
-            exit(EXIT_FAILURE);
-         }
-         KARATSUBA_SQR_CUTOFF = x;
-         t2 = s_time_sqr(x);
-         if ((t2 == 0uLL) || (t2 == UINT64_MAX)) {
-            fprintf(stderr,"Karatsuba squaring failed (%s)\n",
-                    (t2 == 0uLL)?"wrong result":"internal error");
-            exit(EXIT_FAILURE);
-         }
-         if (verbose == 1) {
-            printf("%d: %9"PRIu64" %9"PRIu64", %9"PRIi64"\n", x, t1, t2, (int64_t)t2 - (int64_t)t1);
-         }
-         if (t2 < t1) {
-            if (count == s_stabilization_extra) {
-               count = 0;
-               break;
-            } else if (count < s_stabilization_extra) {
-               count++;
-            }
-         } else if (count > 0) {
-            count--;
-         }
-      }
-      KARATSUBA_SQR_CUTOFF = x - s_stabilization_extra * increment_print;
+      s_run("Karatsuba squaring", s_time_sqr, &KARATSUBA_SQR_CUTOFF);
+      updated.ksqr = KARATSUBA_SQR_CUTOFF;
+      KARATSUBA_SQR_CUTOFF = INT_MAX;
 #endif
 #ifdef BN_S_MP_TOOM_MUL_C
-      if ((verbose == 1) || (testmode == 1)) {
```

Comment:
use MP_HAS instead of all the ifdefs?

Reply:
I did that already and forgot to push, sorry